BACKENDS: Harden grabOverlay in all platforms

lephilousophe opened the pull request:
Use specific copy code where applicable and replace duplicated code by
calls to copyBlit which optimizes blitting

This is a follow-up of 5af11925809267eee1abd5ddf251530f6f5f5572 and 5c8bd3b687ac166ada7d26362aa16523c002d6a6.

crossBlit can't be use on all platforms (PSP and DS especially). Instead of encouraging to pass whatever surface format we want, expect the format to be the one given by `getOverlayFormat`.
Instead of cropping image to surface size and hiding bugs, just make sure we will never overflow allocated memory.
Bug if any must be fixed at the correct place.

This way, all platforms are safe and rendering looks as expected.

ccawley2011 commented:
The changes to use ``copyBlit()`` look good, however I think it would be better to have the calling code handle the size and format checks rather than needing them in every backend.

Regarding @mgerhardy's original issue, I think the problem is that the event recorder isn't correctly allowing ``EVENT_SCREEN_CHANGED`` events through, causing the GUI to be out of sync with the backend when ``grabOverlay()`` is called.

lephilousophe commented:
@ccawley2011 I am not sure about where to put this new function, in graphics/conversion.cpp?
This piece of code is really specific even though it's duplicated across almost all backends.

sev- commented:
To me, those checks look all fine. It makes not much sense of moving those three lines of code to something like `ensureSurfaceIsCouldHoldTheOverlay(w, h, format)` as it is way too specialized.

lephilousophe commented:
I rebased commit and added a new commit which properly fix bug encountered in 5af11925809267eee1abd5ddf251530f6f5f5572

This make sure that this PR won't break the event recorder.

lephilousophe commented:
As no one has anything to say about this PR and that this PR shouldn't break EventRecorder anymore, I merge.
